### PR TITLE
Fix entities restriction on KB items; closes #1391

### LINKED ISSUE #1391: Entities restrictions on KB items
On the KB items list, the restriction seems to be applied correctly.

But on the item itself, access is granted even if we're on another not related (ie. not ancestor nor child) entity.

### PR DESCRIPTION
**Pull request by trasher**
Two problems spotted:

* what is the need for `getSonsOf`?
* recursivity check should not be harcoded.

There are many, many possible use cases on KB along with entities, this would require a big set of unit tests; but probably in the master.

### CODE REVIEW

**Comment by remicollet** [comment]
Can you please the case you are trying to fix ?

BTW same code also used for reminder and rssfeed.

**Comment by trasher** [comment]
The fix concernes issue described in #1391.

Here is the issue:
* create two entities 1, and 2, a child of 1,
* create a KB item, attach it to entity 1 without recursivity,
* create another KB item, attach it to entity 2, without recursivity,

When you set entity 2 as current entity; and you go the the KB items list, you'll see only the second KB item (the one attached to entity 2); that is correct.

But in the same time, if you call the url to display the KB item (`/knowbaseitem.form.php?id=xyz`), you'll see it; which is wrong.
That is what I try to fix here :)

**Comment by trasher** [comment]
I was not aware this code was used elsewhere; but since this is specific to KB; I guess the initial issue is the same there.

**Comment by remicollet** [comment]
Indeed, I can reproduce (and the fix seems ok)

Can you check group check  (~199) and profile check (~439)

More, if a single entity have to be checked, you can switch to haveAccessToEntity.

**Comment by trasher** [comment]
I do not understand the checks you'd want me to add.

On another hand, would not be such changes out of the scpe of the current fix? Should not such changes target master only and not 9.1?

**Comment by remicollet** [comment]
I just mean that this piece of code is used in lot of places, and we should ensure other parts are  ok or also need to be fixed.

And, instead of:

```
               $entities = array($entity['entities_id']);
               if (Session::haveAccessToOneOfEntities($entities, $entity['is_recursive'])) {
                  return true;
               }
```

Simpler:

```
               if (Session::haveAccessToEntity($entity['entities_id'], $entity['is_recursive'])) {
                  return true;
               }
```

And, as this is a bug, 9.1 should be the target.

**Comment by remicollet** [comment]
And, BTW, for master, a try to factorize this stuff between kbitem, rssfeed and reminder could be usefull ;)  and will reduce need for tests.

**Comment by trasher** [comment]
I've made the changes fr the 9.1 branch for now. Once It has been approved, I'll worl on factorization for master.

**Comment by trasher** [comment]
Thanks for the review :)

File: inc/reminder.class.php [hunk 1]
```diff
@@ -217,11 +217,7 @@ function haveVisibilityAccess() {
                      return true;
                   }
                   // Restrict to entities
-                  $entities = array($group['entities_id']);
-                  if ($group['is_recursive']) {
-                     $entities = getSonsOf('glpi_entities', $group['entities_id']);
-                  }
-                  if (Session::haveAccessToOneOfEntities($entities, true)) {
+                  if (Session::haveAccessToEntity($group['entities_id'], $groups['is_recursive'])) {
```
Comment (trasher) [inline inc/reminder.class.php]: There is a typo here :/ Should be `group` not `groups`

**Comment by trasher** [comment]
group(s) typo and missed changed fixed in c6e307c

**Pull request merged**